### Phase command
- [SNP only phasing](#snp-only-phasing)
- [SNP and indel co-phasing](#snp-and-indel-co-phasing)
<!-- - [SNP and SV co-phasing](#snp-and-sv-co-phasing) -->
- [Complete list of phase parameters](#the-complete-list-of-phase-parameters)
- [Phased Genotype Output Format](#phased-genotype-output-format)
- [Output files of Purity](#output-files-of-Purity)
- [Output files of SNP and indel phasing](#output-files-of-snp-and-indel-phasing)
<!-- - [Output files of SNP and SV co-phasing](#output-files-of-snp-and-sv-co-phasing) -->

#### SNP-only phasing
For SNP-only phasing, the input of LongPhase consists of SNPs in VCF (e.g., SNP.vcf), an indexed reference in Fasta (e.g., reference.fasta, reference.fasta.fai), and one (or multiple) indexed read-to-reference alignment in BAM (e.g., alignment1.bam, alignment1.bai, alignment2.bam, ...) (see [Input Preparation](#input-preparation)). An example of SNP phasing usage is shown below.


```bash
longphase-to phase \
-s SNP.vcf \
-b alignment1.bam \
-b alignment2.bam \
-r reference.fasta \
-t 8 \
-o phased_prefix \
--pon-file pon1.vcf,pon2.vcf \
--strict-pon-file pon3.vcf,pon4.vcf
```
#### SNP and indel co-phasing
When the SNP.vcf file contains both SNP and small insertion/deletion variations (indels), use the `--indels` parameter for co-phasing SNPs and indels.
```bash
longphase-to phase \
-s SNP.vcf \
-b alignment.bam \
-r reference.fasta \
-t 8 \
-o phased_prefix \
--indels
```
<!-- #### SNP and SV co-phasing
When co-phasing SNPs and SVs, except for the same input (i.e., SNPs, reference, and alignments), LongPhase takes an extra input of called SVs in VCF (e.g., [SV_sniffles.vcf](https://bioinfo.cs.ccu.edu.tw/bioinfo/download.html)), which should be generated by Sniffles (with --num_reads_report in sniffles1 and --output-rnames in sniffles2) or CuteSV (with --report_readid--genotype)(see [Input Preparation](#input-preparation)).
```bash
longphase-to phase \
-s SNP.vcf \
--sv-file SV.vcf \
-b alignment.bam \
-r reference.fasta \
-t 8 \
-o phased_prefix \
--ont # or --pb for PacBio Hifi
``` -->

#### Complete list of phase parameters
```text
Usage:  phase [OPTION] ... READSFILE
   --help                                 display this help and exit.

require arguments:
   -s, --snp-file=NAME                    input SNP vcf file, or a VCF file with both SNPs and indels.
   -b, --bam-file=NAME                    input bam file.
   -r, --reference=NAME                   reference fasta.   -c, --caller=NAME                      variant caller name.
                                          options: clairs_to_ss, clairs_to_ssrs, deepsomatic_to

optional arguments:
   -t, --threads=Num                      number of thread. default:1
   -o, --out-prefix=NAME                  prefix of phasing result. default: result
   --indels                               phase small indel. default: False
   --dot                                  each contig/chromosome will generate dot file.
   --loh                                  output LOH results. default: False

somatic arguments:
   --purity=[0~1]                         set sample purity directly; if omitted, estimate automatically.
   --disable-calling                      disable longphase calling mode. default: False
   --disable-pon-tag                      disable reading the VCF FILTER field PON to determine germline variants. default: False
   --disable-refine-somatic               do not modify VCF FILTER based on somatic refinement. default: False
   --pon-file=NAME                        input PON VCF file. determines germline variants using position-based matching.
                                          input format: A.vcf,B.vcf
   --strict-pon-file=NAME                 input PON VCF file. determines germline variants using both position and ALT allele matching.
                                          input format: A.vcf,B.vcf
   --somaticConnectAdjacent=Num           connect adjacent N SNPs. default:6

parse alignment arguments:
   -q, --mappingQuality=Num               filter alignment if mapping quality is lower than threshold. default:1
   -x, --mismatchRate=Num                 mark reads as false if mismatchRate of them are higher than threshold. default:3

phasing graph arguments:
   -p, --baseQuality=[0~90]               change edge's weight to --edgeWeight if base quality is lower than the threshold. default:12
   -e, --edgeWeight=[0~1]                 if one of the bases connected by the edge has a quality lower than --baseQuality
                                          its weight is reduced from the normal 1. default:0.1
   -a, --connectAdjacent=Num              connect adjacent N SNPs. default:35
   -d, --distance=Num                     phasing two variant if distance less than threshold. default:300000
   -1, --edgeThreshold=[0~1]              give up SNP-SNP phasing pair if the number of reads of the 
                                          two combinations are similar. default:0.7
   -L, --overlapThreshold=[0~1]           filtering different alignments of the same read if there is overlap. default:0.2 

haplotag read correction arguments:
   -m, --readConfidence=[0.5~1]           The confidence of a read being assigned to any haplotype. default:0.65
   -n, --snpConfidence=[0.5~1]            The confidence of assigning two alleles of a SNP to different haplotypes. default:0.75
```

### Phased Genotype Output Format
---
The **genotype (GT)** uses the `|` delimiter to represent phased alleles on two haplotypes. The format is `A|B`, where:
- `A` represents the allele on the first haplotype
- `B` represents the allele on the second haplotype
- `|` indicates phasing (i.e., the alleles are ordered and assigned to specific haplotypes)

Allele
- `0` reference allele
- `1` alternate allele
- `.` denotes an unknown or undetermined allele

This format is extended with **GT2** and **GT3** fields to enable detailed modeling of somatic variation, phasing complexity, and loss of heterozygosity (LOH) — crucial for applications in cancer genomics.

To facilitate understanding of how different types of variants—such as germline and somatic variants—are represented in phased genotypes, the following two examples are provided:
```
Germline variant GT:GT2:GT3 0|1:./.:./.
Somatic  variant GT:GT2:GT3 0|0:.|1:./.
```

---
#### GT (Germline and Primary Somatic Phasing)
##### 1. Germline heterozygosity
- `0|1`: haplotype 1 is reference; haplotype 2 is alternate
- `1|0`: haplotype 1 is alternate; haplotype 2 is reference

##### 2. Germline homozygosity or LOH
- `.|1`: haplotype 1 unknown (potential LOH); haplotype 2 is alternate
- `1|.`: haplotype 1 is alternate; haplotype 2 unknown (potential LOH)

##### 3. Somatic mutation not in LOH
- `0|0`: both haplotypes are reference

##### 4. Somatic mutation in LOH
- `.|0`: haplotype 1 unknown; haplotype 2 is reference
- `0|.`: haplotype 1 is reference; haplotype 2 unknown

---

#### GT2 and GT3 (Somatic Phasing and LOH Resolution)
These fields provide finer resolution for subclonal or ambiguous somatic phasing, especially in LOH contexts.

##### 1. Somatic mutations not in LOH
- GT2 `.|1`, GT3 `./.`: `hp1-1` unknown; `hp2-1` is alternate (somatic-derived)
- GT2 `1|.`, GT3 `./.`: `hp1-1` is alternate (somatic-derived); `hp2-1` unknown
- GT2 `1|1`, GT3 `./.`: Ambiguous somatic haplotype (`hp3`); both alleles are alternate

##### 2. Somatic mutations in LOH regions
- GT2 `0|.`, GT3 `1|.`: `hp1-1-1` is reference; `hp1-1-2` is alternate
- GT2 `1|.`, GT3 `0|.`: `hp1-1-1` is alternate; `hp1-1-2` is reference
- GT2 `1|.`, GT3 `1|.`: Ambiguous somatic haplotype (`hp3`); both alleles are alternate
- GT2 `.|0`, GT3 `.|1`: `hp1-2-1` is reference; `hp1-2-2` is alternate
- GT2 `.|1`, GT3 `.|0`: `hp1-2-1` is alternate; `hp1-2-2` is reference
- GT2 `.|1`, GT3 `.|1`: Ambiguous somatic haplotype (`hp3`); both alleles are alternate

### Haplotype-Aware Variant Re-Calling Output Format
When performing haplotype-aware variant re-calling, LongPhase refines the VCF FILTER field based on somatic variant analysis. By default, LongPhase may correct non-PASS variants to PASS or revert PASS variants to non-PASS. The `--disable-refine-somatic` option disables this refinement, retaining the original FILTER values. Variants with FILTER value `PASS` indicate they passed LongPhase's somatic variant analysis. LongPhase adds two FILTER tags: `LP_nonSomatic` for non-somatic variants and `LP_PON` for variants found in the panel of normals. For example:
```vcf
##FILTER=<ID=LP_nonSomatic,Description="Non-somatic variant tagged by longphase-to">
##FILTER=<ID=LP_PON,Description="PON variant tagged by longphase-to">
#CHROM  POS     ID      REF     ALT     QUAL    FILTER        INFO    FORMAT  SAMPLE
chr1    10864   .       G       A       23.2667 PASS          ...     ...     ...
chr1    14695   .       C       T       17.3507 LP_nonSomatic ...     ...     ...
chr1    34932   .       T       C       2.8045  LP_PON        ...     ...     ...
```

If the `--disable-refine-somatic` option is used, variants can be filtered using the following command:
```bash
awk '/^#/ || /0\|0/ || /\.|0/ || /0\|./' <vcf_path> > <output_path>
```

### Output files of Purity
When performing purity estimation, LongPhase outputs the tumor purity estimate in the VCF header as a double. It uses the tag `tumor_purity`. For example:
```vcf
##tumor_purity=0.6
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  SAMPLE
```


### Output files of SNP and indel phasing
When phasing SNPs alone, LongPhase outputs the results into a VCF file. The alleles of the two haplotypes are stored in the GT field (e.g., 1|0), whereas the left and right alleles of the vertical bar represent the paternal or maternal haplotypes. The last PS field (e.g., 16809) represents the identifier of the block. For instance, the following example illustrates two haplotypes of five-phased SNPs, CCCCC and GATGT, in the same block 16809. The output of phased indels is similar and in the same VCF.

```vcf
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##FORMAT=<ID=PS,Number=1,Type=Integer,Description="Phase set identifier">
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  default
1       16809   .       C       G       8.4     PASS    .       GT:GQ:DP:AD:VAF:PL:PS   1|0:8:79:51,28:0.35443:7,0,16:16809
1       16949   .       A       C       8.4     PASS    .       GT:GQ:DP:AD:VAF:PL:PS   0|1:8:67:43,21:0.313433:7,0,25:16809
1       21580   .       C       T       13.9    PASS    .       GT:GQ:DP:AD:VAF:PL:PS   1|0:14:75:50,24:0.32:13,0,30:16809
1       23359   .       C       G       13.2    PASS    .       GT:GQ:DP:AD:VAF:PL:PS   1|0:13:52:24,18:0.346154:13,0,42:16809
1       24132   .       C       T       11.1    PASS    .       GT:GQ:DP:AD:VAF:PL:PS   1|0:11:63:41,17:0.269841:10,0,29:16809
```

<!-- ### Output files of SNP and SV co-phasing
When co-phasing SNPs and SVs, two VCFs (one for SNPs and one for SVs) are outputted. Similarly, the phased SVs are stored in the GT field and the block ID is in the PS field. For instance, the following example illustrates two haplotypes of five SNPs and two SVs, A\<INS\>G\<noSV\>TCC and G\<noSV\>A\<INS\>ATT, which are co-phased in the same block 382189.

An example of SNP VCF file
```vcf
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##FORMAT=<ID=PS,Number=1,Type=Integer,Description="Phase set identifier">
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  default
1       465289  .       A       G       34.2    PASS    .       GT:GQ:DP:AD:VAF:PL:PS   1|0:4:16:2,11:0.6875:31,0,1:382189
1       544890  .       G       A       3.1     PASS    .       GT:GQ:DP:AD:VAF:PL:PS   1|0:3:31:30,0:0:0,0,31:382189
1       545612  .       T       A       6.8     PASS    .       GT:GQ:DP:AD:VAF:PL:PS   1|0:7:37:28,8:0.216216:5,0,32:382189
1       545653  .       C       T       14      PASS    .       GT:GQ:DP:AD:VAF:PL:PS   1|0:14:44:30,14:0.318182:13,0,26:382189
1       561458  .       C       T       5.1     PASS    .       GT:GQ:DP:AD:VAF:PL:PS   1|0:5:17:14,0:0:3,0,17:382189
```
An example of SV VCF file
```vcf
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##FORMAT=<ID=PS,Number=1,Type=Integer,Description="Phase set identifier">
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  default
1       534057  7       N       AAATTCGCGCATATCACGGGTGCCGCCTCTGTGCAGCTCACGAAACGCCATACTACGGTGCTGCTCAGCAGCTACGGAATCGCTATACCTACGCGAGCTGCCTCAGCAGCCAC       .       PASS    IMPRECISE;SVMETHOD=Snifflesv1.0.11;CHR2=1;END=534128;STD_quant_start=25.369041;STD_quant_stop=29.191054;Kurtosis_quant_start=-0.239012;Kurtosis_quant_stop=-1.333733;SVTYPE=INS;RNAMES=0120d560-50f0-4298-8b03-7bd30f3cf139,030ac5d4-e616-4ce9-8ad3-243835335085,0cf1b0d9-2b4d-463d-a658-01b4b040dc63,1c9e982a-8af7-4ba0-8cc2-154a679c72e2,22e11f79-0067-4735-8b69-97d951ca702f,2ca8a6f4-be9d-4df5-80d2-dc1743f97a84,35dff960-22b6-4216-af69-8878b8860362,390d2fb4-9224-41a1-a9fe-6cb3bbe4273a,3e333422-12ca-4f16-afb8-ed7611dcbc2c,3e8ed78a-b857-4941-bbc1-52ca51e26c08,4191371c-49ea-466d-aadc-06f27cdf1050,4aaae789-54fe-4fa5-84b3-5524dc2b3796,581e0cfb-2491-44d7-a2e1-ba1516ba0f2f,59749531-9abf-4ff4-a4a1-31484ba3d32d,5c97b0a9-925e-4153-952d-0f437171d3dc,6067590e-956c-442f-bbb7-cae597d616ad,623804bb-e2fe-415d-96ae-3d06aec63e5d,672244ce-2d5d-45cf-beb2-ddeddae917e8,6b79aa23-7c9c-49dc-9b88-8419c88c7a36,6e60d235-6654-4ef8-9feb-70f12a397721,6fbb55c5-57fc-43bc-8a24-b0058778054c,8e10bf13-9674-489c-924e-182a42e08a34,aa6ba092-4221-4d54-8819-811448c34983,af2169b3-b308-4db5-9675-15ff5f68d8dd,b214fcbd-77de-4dd5-84db-6d2b7e1f3158,c140eaba-e0e7-44e7-9f16-c8c67fd4a2f2,c7835cf7-44c0-44da-b10e-b2468fc8caab,ca4aa84d-34d1-4639-8634-b6a5540129ca,caba4bde-cdc5-4344-9803-a3c158525b0c,e0747feb-60bb-40db-a144-a9b43dd13256,e6992c7d-c00e-40e7-b80b-562094a9b60f,e8bb376c-20e0-4bed-a61f-b82b5c37ef6f,f3242a61-deec-49e7-b99f-335a1ba13791,f87dfdf7-7b68-421b-b395-3769a5fa3ac1,f91a7627-7fdb-4f03-8f33-0ed1649d96fe;SUPTYPE=AL;SVLEN=43;STRANDS=+-;RE=35;REF_strand=44;AF=0.795455    GT:DR:DV:PS     0|1:9:35:382189
1       545892  8       N       ACACGCGGGCCGTGGCCAGCAGGCGGCGCTGCAGGAGAGGAGATGCCCAGGCCTGGCGGCC   .       PASS    IMPRECISE;SVMETHOD=Snifflesv1.0.11;CHR2=1;END=545893;STD_quant_start=28.919840;STD_quant_stop=28.543200;Kurtosis_quant_start=-0.382251;Kurtosis_quant_stop=-0.130808;SVTYPE=INS;RNAMES=0120d560-50f0-4298-8b03-7bd30f3cf139,030ac5d4-e616-4ce9-8ad3-243835335085,0cf1b0d9-2b4d-463d-a658-01b4b040dc63,22e11f79-0067-4735-8b69-97d951ca702f,2ca8a6f4-be9d-4df5-80d2-dc1743f97a84,3977c988-9901-4e5b-9f9c-b8ebfcce8e93,3e333422-12ca-4f16-afb8-ed7611dcbc2c,4191371c-49ea-466d-aadc-06f27cdf1050,4aaae789-54fe-4fa5-84b3-5524dc2b3796,5933e1b7-1aeb-4437-a875-3befbf703420,623804bb-e2fe-415d-96ae-3d06aec63e5d,672244ce-2d5d-45cf-beb2-ddeddae917e8,6b79aa23-7c9c-49dc-9b88-8419c88c7a36,7842d9f1-9a77-4c9a-ab5b-5a644ed2d355,7ba26d64-d9b0-475f-8d5f-1fa73fc42d93,8e10bf13-9674-489c-924e-182a42e08a34,a2b1b2ef-1e28-465e-8b3f-c44e15990d8b,a45514f1-4aae-40eb-94eb-2969722a7b05,b8181546-6839-49cd-b64f-b65c96369a2b,c140eaba-e0e7-44e7-9f16-c8c67fd4a2f2,c7835cf7-44c0-44da-b10e-b2468fc8caab,ca4aa84d-34d1-4639-8634-b6a5540129ca,d56f0abe-4389-4197-a151-0eb567fb99f0,e6992c7d-c00e-40e7-b80b-562094a9b60f,e8bb376c-20e0-4bed-a61f-b82b5c37ef6f,ec325153-0c55-4ece-8f3c-c432701e6750,f3242a61-deec-49e7-b99f-335a1ba13791,f91a7627-7fdb-4f03-8f33-0ed1649d96fe;SUPTYPE=AL;SVLEN=62;STRANDS=+-;RE=28;REF_strand=51;AF=0.54902        GT:DR:DV:PS     1|0:23:28:382189
``` -->
